src/CanvasInput/index.js

```diff
@@ -59,47 +53,40 @@ export default class CanvasInput extends React.Component {
     }
   }
 
-  handleClick = evt => {
-    evt.stopPropagation()
-    evt.preventDefault()
-    if (!this.state.mouseDrag) {
-      const mouse = getMouseFromEvent(evt)
-      const solvedData = runHoverSolverOn(
-        getDataUnderMouse(this.props, mouse, this.canvasRef.current)
-      )
-      this.props.onUpdate({
-        action: 'mouseClick',
-        mouse,
-        renderDatum: solvedData.renderDatum,
-        hoverRenderData: solvedData.hoverRenderData,
-        hoverData: solvedData.hoverData,
-        localMouse: solvedData.localMouse,
-        layerProps: solvedData.layerProps,
-        rootProps: this.props.rootProps,
-      })
-    }
-    this.setState({mouseDrag: false})
+  handleClick = event => {
+    event.stopPropagation()
+    event.preventDefault()
+    event.persist()
+    this.setState(prevState => {
+      if (!prevState.mouseDrag) {
+        const mouse = getMouseFromEvent(event)
+        const solvedData = runHoverSolverOn(
+          getDataUnderMouse(this.props, mouse, this.canvasRef.current)
+        )
+        this.props.onUpdate({
+          action: 'mouseClick',
+          mouse,
+          ...solvedData,
+          rootProps: this.props.rootProps,
+        })
```

Comment:
AFAIK it's unsafe for a `setState` updater to have side effects. However it's not unsafe to access `this.state` in an event handler (as long as it's not inside of an updater), so I think the previous iteration was correct.